### DISCUSSION TITLE
Redirection and spinner

### DISCUSSION THREAD

**Discussion by EmilyGraceSeville7cf** [charmbracelet/gum#527] [Q&A]
How to run command with redirection with `gum spin`? I have these functions:

```fish
function spin --argument-names title
    set --local color '#1d87da'
    set --erase argv[1]
    gum spin --title.foreground=$color \
        --spinner.foreground=$color \
        --show-output \
        --title=$title -- $argv
end

function to_file --argument-names source output
    jsdoc2md $source >>$output
end
```

I wanna call `spin` like this: `spin "Generating documentation for $file..."  to_file $file $output`, but when I do I get all my Markdown documentation discarded. How do I fix this issue?

Full code:

```fish
#!/usr/bin/env fish

function error
    echo (set_color red)$argv(set_color normal)
end

function info
    echo (set_color green)$argv(set_color normal)
end

function confirm
    set --local color '#1d87da'
    gum confirm --prompt.foreground=$color \
        --selected.foreground='#e5f6fd' \
        --selected.background=$color -- $argv
end

function spin --argument-names title
    set --local color '#1d87da'
    set --erase argv[1]
    gum spin --title.foreground=$color \
        --spinner.foreground=$color \
        --show-output \
        --title=$title -- $argv
end

function to_file --argument-names source output
    jsdoc2md $source >>$output
end

set generator jsdoc2md

type --query $generator || begin
    error "No '$generator' package found to generate documentation."
    if confirm "Do you want to install '$generator' package?"
        spin "Installing '$generator' package..." npm install --global jsdoc-to-markdown
    else
        info Installation cancelled
        exit
    end
end

set output Documentation.md
set ignored_files ActionInfo_.js SimpleValidator.js ComplexValidator.js

echo >$output
for file in *.js
    contains $file $ignored_files && begin
        error "$file is ignored"
        continue
    end
    spin "Generating documentation for $file..." to_file $file $output # <-- Problem here
end
```

**Comment by wellkilo** [comment]
There are two shell-boundary issues here:

1. gum spin executes argv directly; it does not ask Fish to parse redirection operators.
2. to_file is a Fish function in the parent shell, not an executable that the child process can launch.

Run a shell explicitly and perform the redirection in that shell:

    function spin_to_file --argument-names title source output
        set --local color '#1d87da'
        gum spin \
            --title.foreground=$color \
            --spinner.foreground=$color \
            --title=$title -- \
            fish -c 'jsdoc2md $argv[1] >> $argv[2]' -- $source $output
    end

    spin_to_file "Generating documentation for $file..." $file $output

Quote the variables in your real script if filenames can contain spaces; Fish preserves each argument passed after the -- separator as an argv element.

Also remove --show-output for this operation. You are intentionally redirecting jsdoc2md's stdout to the file, so there is no stdout left for gum to display. stderr will still determine/report failure; on current Gum you can use --show-error if you only want captured output printed when the command fails.

The reason is visible in Gum's [spin implementation](https://github.com/charmbracelet/gum/blob/7179388031ae67d7f538d001be87d931f1cf5e28/spin/spin.go#L50-L61): it calls exec.CommandContext(command[0], command[1:]...), so shell functions and operators such as >> only work when you explicitly invoke a shell.
